### ISSUE TITLE
Cannot import 'electron' module in Frontend Project

### ISSUE THREAD

**Issue by ComradeCow** [bennymeg/nx-electron#43]
**Describe the bug**
I want to use Electron's IPC system to communicate between the Main and Renderer processes.

Specifically I get this error:
> ERROR in C:/git/webrtc-play/node_modules/electron/index.js
Module not found: Error: Can't resolve 'fs' in 'C:\git\webrtc-play\node_modules\electron'

**To Reproduce**
* Generate front-end application (eg `nx g @nrwl/react:app my-react-app`)
* Generate electron (`nx g nx-electron:application electron-shell --frontendProject my-react-app`)
* Start front-end `nx serve my-react-app`
* Start electron shell `nx serve electron-shell`

Add import in React app:

```ts
import { ipcRenderer } from 'electron';
```

**Expected behavior**
I expected to be able to import the electron module in my frontend application.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Windows 10
 - Nx Electron Version: 9.3.2

**Additional context**

I'm not 100% certain if the the set up I have is correct - but by looking at code, I assume the intent is to use `nx serve` to host the front-end app, then the electron app just points the browser at `http://localhost:4200`. If that's the case, do we need some way to change the [webpack target](https://webpack.js.org/configuration/target/) to `electron-renderer` to facilitate all that?  I tried using a `webpackConfig` script to swap out the config.  That fixed the imports, but I'm not receiving messages as expected (and seems _suuuper_ hacky)

workspace.json:
```jsonc
{
  // snipped for clairity
  "webrtc-playground": {
    "architect": {
       "build": {
           "webpackConfig": "electron-host-webpack.config"
        } 
     } 
  }
}
```

electron-host-webpack.config
```js
const getReactWebpackConfig = require('@nrwl/react/plugins/webpack');

function getWebpackConfig(config /*: Configuration*/) {
  config = getReactWebpackConfig(config);

  config.target = 'electron-renderer';

  return config;
}

module.exports = getWebpackConfig;
```

Is this user error? Are there suggestion for how to handle communication between the renderer and main threads?

**Comment by bennymeg** [comment]
I think this is a react related issue.
You can't import electron directly in react / angular.
Read this mini [article](https://erikmartinjordan.com/electron-react-ipcrenderer) about importing electron in react.
For angular, use [ngx-electron](https://www.npmjs.com/package/ngx-electron).

**Comment by esakal** [comment]
@ComradeCow I actually had this issue yesterday :) to resolve it you must customise the webpack configuration that builds your react application with `target = 'electron-renderer'`.

I will add here steps how to do it soon.

**Comment by esakal** [comment]
@ComradeCow continuing my previous comment:
1. open `workspace.json` file, and find property `webpackConfig` under your **React** application configuration.
2. Replace property  `webpackConfig` with value `tools/webpack-react-electron.config.js`.
3. create file `tools/webpack-react-electron.config.js`.
4. copy the following into that file
```
module.exports = (config, options) => {

  let nrwlConfig = require("@nrwl/react/plugins/webpack")(config);

  return {
    ...nrwlConfig,
    target: 'electron-renderer'
  }
}
```
5. if you are already running nrwl serve, stop it

That's it, next time you serve the React app it should work for you.

I hope this helps.
Eran

**Comment by ComradeCow** [comment]
@bennymeg  Thanks for the tips - but they haven't worked for me, unfortuantely.

Following the link you posted, I created a preload to get the ipcRenderer hooked in.
```js
const { ipcRenderer } = require('electron');
window.ipcRenderer = ipcRenderer;
```
in [electron project]/src/app/app.ts
```ts
private static initMainWindow() {
 webPreferences: {
    preload: path.join(__dirname, 'assets', 'preload.js'),
  }
}
```
(btw @esakal  - using that preload script worked to get the render in, I didn't need the webpackConfig workaround anymore)

As far as my events are hooked up (to hopefully rule out user error)

In `electron.events.ts`
```ts
static bootstrapElectronEvents() {
    console.log('hooking up ping event');
    ipcMain.on('ping-me', (event) => {
      console.log('received ping!');

      ipcMain.emit('pong-you');
    })
}
```

In my frontend project's `app.tsx`
```ts
export const App = (props) => {
  useEffect(() => {
    console.log('hooking up pong event');

    window.ipcRenderer.on('pong-you', () => {
      console.log('I got a pong!');
    });

    window.ipcRenderer.emit('ping-me');
  }, []);
  /* snipping for clairity */
}
```
`ipcRenderer` is defined for the browser, but when I debug into it, there's no event handler for `ping-me` registered. So EventEmitter doesn't emit anything (returns false). I don't know much about Electron, but I had expected to see _something_ to bridge communication back to the main process.

Any other ideas for me to try?

**Comment by bennymeg** [comment]
Hmm, I am not very experienced with react but that seems about right to me.

It is probably better to wait for an react developer to answer that (Eran?), but I can try to help you debug it.
- Do you see the `'hooking up pong event'` & `'hooking up pong event'` log message?
- Can you try using [ipcRenderer.invoke()](https://www.electronjs.org/docs/api/ipc-renderer#ipcrendererinvokechannel-args) & [ipcMain.handle()](https://www.electronjs.org/docs/api/ipc-main#ipcmainhandlechannel-listener)?

Anyway, it does not seem like an issue in `nx-electron`...

**Comment by ComradeCow** [comment]
I was finally able to get back and look into this more - this doesn't seem nx-electron specific.  I fired up a vanilla electron app to see what happens, and I ran into similar issues.  Seems like `'on'` and `'emit'` shouldn't be used in favor of `'handle'` and `'invoke'` or `'send'`.  I'm still really hazy on it, so I need to do more research.   

I'm not certain how important that `webpack-renderer` target is - but hopefully the sample code here will someone coming in from search results.

Closing.
